fix: add note for drain stuck when upgrade from v1.4.0 to v1.4.1 (#707)
harvester/docs: docs/upgrade/v1-4-0-to-v1-4-1.md

---
sidebar_position: 2
sidebar_label: Upgrade from v1.4.0 to v1.4.1
title: "Upgrade from v1.4.0 to v1.4.1"
---

<head>
  <link rel="canonical" href="https://docs.harvesterhci.io/v1.4/upgrade/v1-4-0-to-v1-4-1"/>
</head>

## General information

An **Upgrade** button appears on the **Dashboard** screen whenever a new Harvester version that you can upgrade to becomes available. For more information, see [Start an upgrade](./automatic.md#start-an-upgrade).

For air-gapped environments, see [Prepare an air-gapped upgrade](./automatic.md#prepare-an-air-gapped-upgrade).

## Known issues

---

> ### 1. An upgrade is stuck in the Pre-drained state due to Longhorn orphan engine(s)
> 
> You might see an upgrade is stuck in the "pre-drained" state:

Reviewer: ```suggestion
### 1. Upgrade is stuck in the "Pre-drained" state

The upgrade process may become stuck in the "Pre-drained" state. Kubernetes is supposed to drain the workload on the node, but some factors may cause the process to stall.
```

![](/img/v1.2/upgrade/known_issues/3730-stuck.png)

> In this stage, Kubernetes is supposed to drain the workload on the node, but some reasons might cause the process to stall.
> 
> Longhron instance manager's orphan engines process(es) could cause this issue. To check if that's the case, perform the following steps:

Reviewer: ```suggestion
A possible cause is processes related to orphan engines of the Longhorn Instance Manager. To determine if this applies to your situation, perform the following steps:
```

> 1. Assume the stuck node is `harvester-node-1`.
> 1. Check the `instance-manager` pod's name on the stuck node:

Reviewer: ```suggestion
1. Check the name of the `instance-manager` pod on the stuck node.

    Example:

    The stuck node is `harvester-node-1`, and the name of the Instance Manager pod is `instance-manager-d80e13f520e7b952f4b7593fc1883e2a`.
```

    ```
    $ kubectl get pods -n longhorn-system --field-selector spec.nodeName=harvester-node-1 | grep instance-manager
    instance-manager-d80e13f520e7b952f4b7593fc1883e2a          1/1     Running   0              3d8h
    ```

>     The output above shows that the `instance-manager-d80e13f520e7b952f4b7593fc1883e2a` pod is on the node.
> 
> 1. Check Longhorn manager logs and verify that the `instance-manager` pod can't be drained (because of engine `pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0`):

Reviewer: ```suggestion
1. Check the Longhorn Manager logs for informational messages.

    Example:
```

    ```
    $ kubectl -n longhorn-system logs daemonsets/longhorn-manager
    ...
    time="2025-01-14T00:00:01Z" level=info msg="Node instance-manager-d80e13f520e7b952f4b7593fc1883e2a is marked unschedulable but removing harvester-node-1 PDB is blocked: some volumes are still attached InstanceEngines count 1 pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0" func="controller.(*InstanceManagerController).syncInstanceManagerPDB" file="instance_manager_controller.go:823" controller=longhorn-instance-manager node=harvester-node-1
    ```

> 1. Run the command to check if the engine is still running on the stuck node:

Reviewer: ```suggestion
    The `instance-manager` pod cannot be drained because of the engine `pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0`.

1. Check if the engine is still running on the stuck node.
    
    Example:
```

    ```
    $ kubectl -n longhorn-system get engines.longhorn.io pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0 -o jsonpath='{"Current state: "}{.status.currentState}{"\nNode ID: "}{.spec.nodeID}{"\n"}'
    Current state: stopped
    Node ID:
    ```

>     If the output shows the engine is not running, then we can conclude this issue happens.
> 
> 1. Before applying the workaround, check if all volumes are healthy:

Reviewer: ```suggestion
    The issue likely exists if the output shows that the engine is not running.

1. Check if all volumes are healthy.
```

    ```
    kubectl get volumes -n longhorn-system -o yaml | yq '.items[] | select(.status.state == "attached")| .status.robustness'
    ```

>     The output should all be `healthy`. If this is not the case, you might want to uncordon nodes to make the volume healthy again.

Reviewer 2: Is there another way to address this situation? I feel uncordoning the node is not a good idea, and Rancher will soon cordon it again.

Author: Uncord is to let the IM pod be re-created to schedule replicas and make all volumes healthy (if not). In fact, the upgrade controller should ensure all the volumes are healthy in the pre-drain state. 

I borrow the idea from https://docs.harvesterhci.io/v1.4/upgrade/v1-1-2-to-v1-2-0#52-misconfigured-longhorn-instance-manager-r-pod-disruption-budgets-pdb

Hi @Vicente-Cheng,
Can you help to check if I missed anything? Thanks

Author: Removed the uncordon part, based on discussion with @Vicente-Cheng, thanks.

>     The output should all be `healthy`. If this is not the case, you might want to uncordon nodes to make the volume healthy again.

Reviewer: ```suggestion
    All volumes must be marked `healthy`. If this is not the case, you can uncordon nodes to make volumes healthy again.
```

> 1.  Remove the instance manager's PDB:

Reviewer: ```suggestion
1. Remove the `instance-manager` pod's PodDisruptionBudget (PDB) .

    Example:
```

    ```
    kubectl delete pdb instance-manager-d80e13f520e7b952f4b7593fc1883e2a -n longhorn-system
    ```

> - Related issue:

Reviewer: ```suggestion
Related issues:
```

---
sidebar_position: 2
sidebar_label: Upgrade from v1.4.0 to v1.4.1
title: "Upgrade from v1.4.0 to v1.4.1"
---

<head>
  <link rel="canonical" href="https://docs.harvesterhci.io/v1.4/upgrade/v1-4-0-to-v1-4-1"/>
</head>

## General information

An **Upgrade** button appears on the **Dashboard** screen whenever a new Harvester version that you can upgrade to becomes available. For more information, see [Start an upgrade](./automatic.md#start-an-upgrade).

For air-gapped environments, see [Prepare an air-gapped upgrade](./automatic.md#prepare-an-air-gapped-upgrade).

## Known issues

---

### 1. Upgrade is stuck in the "Pre-drained" state

The upgrade process may become stuck in the "Pre-drained" state. Kubernetes is supposed to drain the workload on the node, but some factors may cause the process to stall.

![](/img/v1.2/upgrade/known_issues/3730-stuck.png)

> A possible cause is processes related to orphan engines of the Longhorn Instance Manager. To determine if this applies to your situation, perform the following steps:

Reviewer 3: Should we add the logs showing that the drain is block by the PDB of Longhorn instance manager? This will make sure that the issue is in fact related to Longhorn Instance Manager only

Reviewer 4: The example under item 2 below includes this log line:

`time="2025-01-14T00:00:01Z" level=info msg="Node instance-manager-d80e13f520e7b952f4b7593fc1883e2a is marked unschedulable but removing harvester-node-1 PDB is blocked: some volumes are still attached InstanceEngines count 1 pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0" func="controller.(*InstanceManagerController).syncInstanceManagerPDB" file="instance_manager_controller.go:823" controller=longhorn-instance-manager node=harvester-node-1`

That does mention the PDB block. Is this sufficient, or do we need more log information (or explanation)?

Author: Thanks for the explanation from @tserong, 

@PhanLe1010 IMO, the LH manager's message should be sufficient to make sure the node drain stuck is from LH PDB

Reviewer 3: I agree @tserong @WebberHuang1118 !

> 1. Check the name of the `instance-manager` pod on the stuck node.

Reviewer 5: Could we add another log checking?
I remember that we could find the stuck pod name on the logs of rancher pod.
(Like this https://docs.harvesterhci.io/v1.4/upgrade/v1-1-2-to-v1-2-0/#5-an-upgrade-is-stuck-in-the-pre-drained-state, in `3. Check Rancher logs and verify that the instance-manager-e or instance-manager-r pod can't be drained`)

We can find the pod name and compare it with the command output from the following. To ensure the pod is what we want to handle (stuck).
WDYT?

Author: As described in https://github.com/harvester/docs/blob/d90daa318a37b64718db686833d5ebbc1df41f3f/docs/upgrade/v1-4-0-to-v1-4-1.md#L41-L51

Since LH `v1.7.1` and `v1.6.3`, LH details why the PDB deletion is not allowed https://github.com/longhorn/longhorn-manager/pull/3045. IIUC, this should be enough to find the related IM pod?

Reviewer 5: Sure, only the im-pod logs also work. Thanks!

    instance-manager-d80e13f520e7b952f4b7593fc1883e2a          1/1     Running   0              3d8h
    ```

1. Check the Longhorn Manager logs for informational messages.

    Example:

    ```
    $ kubectl -n longhorn-system logs daemonsets/longhorn-manager
    ...
    time="2025-01-14T00:00:01Z" level=info msg="Node instance-manager-d80e13f520e7b952f4b7593fc1883e2a is marked unschedulable but removing harvester-node-1 PDB is blocked: some volumes are still attached InstanceEngines count 1 pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0" func="controller.(*InstanceManagerController).syncInstanceManagerPDB" file="instance_manager_controller.go:823" controller=longhorn-instance-manager node=harvester-node-1
    ```

    The `instance-manager` pod cannot be drained because of the engine `pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0`.

1. Check if the engine is still running on the stuck node.
    
    Example:

    ```
    $ kubectl -n longhorn-system get engines.longhorn.io pvc-9ae0e9a5-a630-4f0c-98cc-b14893c74f9e-e-0 -o jsonpath='{"Current state: "}{.status.currentState}{"\nNode ID: "}{.spec.nodeID}{"\n"}'
    Current state: stopped
    Node ID:
    ```

>     The issue likely exists if the output shows that the engine is not running.

Reviewer 3: Can we also add the case that engine CR is not found at all?

Reviewer 4: What do we need to do in that case? If the engine is not found at all do we just skip to step 5 and remove the PDB?

Sorry if I I don't know enough about the internals here, I'm just trying to shepherd this docs PR along :-)

Author: Updated per suggestion, thanks.

Reviewer 3: > What do we need to do in that case? If the engine is not found at all do we just skip to step 5 and remove the PDB?

@tserong the current version that @WebberHuang1118 pushed looks good
